pkg/license/license.go

```diff
@@ -0,0 +1,329 @@
+					Path:   filePath,
+					Weight: weight,
+				})
+				break
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the license files by their 'score' (weight)
+	sort.Slice(licenseFiles, func(i, j int) bool {
+		return licenseFiles[i].Weight > licenseFiles[j].Weight
+	})
+
+	return licenseFiles, nil
+}
+
+// melangeLicenseToLicense is a helper that converts a melange license info to a License struct.
+func melangeLicenseToLicense(license string, licensePath string) License {
+	return License{
+		Name:       license,
+		Type:       "",
+		Confidence: 0.0,
+		Source:     licensePath,
+	}
+}
+
+// LicenseCheck checks the licenses of the files in the given filesystem against the melange configuration.
+func LicenseCheck(ctx context.Context, cfg *config.Configuration, fsys fs.FS) ([]LicenseDiff, error) {
+	log := clog.FromContext(ctx)
+
+	// Find all license-text files
+	licenseFiles, err := FindLicenseFiles(fsys)
+	if err != nil {
+		return nil, fmt.Errorf("finding license files: %w", err)
+	}
+
+	if len(licenseFiles) == 0 {
+		// No license files detected, no linting performed.
+		return nil, nil
+	}
+
+	classifier, err := NewClassifier()
+	if err != nil {
+		return nil, fmt.Errorf("creating classifier: %w", err)
+	}
+
+	melangeClassifier := classifier.(*melangeClassifier)
+
+	log.Infof("detected the following licenses in the source code:")
+	detectedLicenses := []License{}
+	for _, lf := range licenseFiles {
+		dl, err := melangeClassifier.Identify(fsys, lf.Path)
+		if err != nil {
+			return nil, fmt.Errorf("identifying license: %w", err)
+		}
+
+		// Print out the licensing information
+		for _, l := range dl {
+			s := ""
+			// This is heuristics, but we want to ignore licenses with a confidence lower than a threshold
+			if l.Confidence < 0.9 {
+				s = " ignored"
+			}
+			log.Infof("  %s: %s (%f%s)", l.Source, l.Name, l.Confidence, s)
+
+			if s == "" {
+				detectedLicenses = append(detectedLicenses, l)
+			}
+		}
+	}
+
+	log.Infof("checking gathered license information against the configuration")
+
+	// Let's first turn the melange licensing information into a coherent licensing list, similar to what Identify returns
+	// We first start off by splitting license information that has OR and AND into separate license entries
+	// Every entry can have multiple AND or ORs
+	melangeLicenses := []License{}
+	for _, ml := range cfg.Package.Copyright {
+		if strings.Contains(ml.License, " OR ") || strings.Contains(ml.License, " AND ") {
+			// Split the license into separate entries using regexp
+			splitLicenses := regexp.MustCompile(`\s+( AND | OR )\s+`).Split(ml.License, -1)
+			for _, sl := range splitLicenses {
+				melangeLicenses = append(melangeLicenses, melangeLicenseToLicense(sl, ml.LicensePath))
+			}
+		} else {
+			melangeLicenses = append(melangeLicenses, melangeLicenseToLicense(ml.License, ml.LicensePath))
+		}
+	}
+
+	// Now let's check if the detected licenses are in the configuration
```

Comment:
this feels like a possible place to split this function in 2 for readability

Reply:
Excellent idea! Even split it into 3